content/blog/gotchaR-bumps.md

```diff
@@ -0,0 +1,162 @@
+and some dabblings with assemblers of various breeds. Mostly, my adjustment has been
+about doing things the way R likes; turning loop-ish code into vector-ish code, using
+the right base functions properly and thinking about data in columns for a start.
+
+Occasionally though, R has the capacity to make me doubt either its sanity, or
+my own, at least for a while. Here are a couple of my lowest points. More may come later, but 
+two is a start, and those PRs for [Hacktoberfest](https://hacktoberfest.digitalocean.com/) 
+aren't going to happen on their own...
+
+# The dodgy dollar?
+
+Here is a piece of code, in which I incrementally ask my friends what their favourite
+fruit is. I decided to use generic variables names, so that later on, I can maybe ask them
+what's their favourite vegetable, or pet, or something. This sort of knowledge, I feel, is 
+essential for being a good considerate colleague. I start writing some code to process
+my friends' replies:-
+
+```
+data <- list()
+data$thing_type <- "Fruit"
+
+for (f in c("Apple", "Banana", "Cherry")) {
+  data$thing <- c(data$thing, f)
+}
+```
+
+So, now I can head to Tescos with a list of my colleague's favourite fruits.
+
+```
+> data$thing
+[1] "Fruit"  "Apple"  "Banana" "Cherry"
+```
+
+Except... what's with the "Fruit" entry then? How did that get into my list of things? Even 
+though I don't think I should need it, I instinctively put `data$thing <- NULL` as the 
+second line. It makes no difference.
+
+I restart RStudio, reboot my computer, update R - the usual sort of random flailing thing. 
+All useless. I start to wonder if R has been tricking me all along, and I do some really
+basic assignments and concatenations to check it's not been fatally flawed all these years. 
+It appears fundamentally sound. 
+
+Eventually, I ask my more R-experienced colleagues if, in fact, I am going insane. They
+cannot answer this for sure, but looking at the programming issue, they tell me it is a 
+feature of R in a way. 
+
+The first time I read from `data$thing`, it doesn't exist 
+in the list, and R lexically auto-completes it into `data$thing_type` - hence the 
+"Fruit". It doesn't auto-complete on the assignment side though, so after the first
+loop iteration, `data$thing` exists, and things continue normally. 
+
+It's tempting to think
+`data$thing <- NULL` beforehand would help - but with a list, that
+would only remove the (non-existent) `thing` from list, because that's what NULL assignment
+with lists does.
+
+So what should I instead do? One option is to replace my inner loop with
+` data$thing <- c(data[['thing']], f)`, because the auto-completion is activated on the 
+use of the dollar sign, but not on a quoted variable name. Moreover, setting these options:
+```
+warnPartialMatchAttr = TRUE
+warnPartialMatchDollar = TRUE
+warnPartialMatchArgs = TRUE
+``` in `.Rprofile` will throw a warning every time R does a partial-match. This seems like
```

Comment:
newline after the closing triple-backtick to avoid formatting issues

Reply:
Done